portal-impl/test/integration/com/liferay/portal/integration/point/PortalIntegrationPointTest.java

```diff
@@ -0,0 +1,205 @@
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.integration.point;
+
+import aQute.bnd.osgi.Builder;
+import aQute.bnd.osgi.Jar;
+
+import com.liferay.portal.integration.point.bundle.portalintegrationpoint.TestDisplayContextFactory;
+import com.liferay.portal.kernel.display.context.DisplayContextFactory;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.io.unsync.UnsyncByteArrayInputStream;
+import com.liferay.portal.kernel.io.unsync.UnsyncByteArrayOutputStream;
+import com.liferay.portal.kernel.util.ServiceProxyFactory;
+import com.liferay.portal.module.framework.ModuleFrameworkUtilAdapter;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+import com.liferay.registry.Registry;
+import com.liferay.registry.RegistryUtil;
+import com.liferay.registry.ServiceReference;
+import com.liferay.registry.ServiceTracker;
+import com.liferay.registry.ServiceTrackerCustomizer;
+
+import java.io.File;
+import java.io.InputStream;
+
+import java.net.URL;
+
+import java.util.Properties;
+
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * @author Leon Chi
+ */
+public class PortalIntegrationPointTest {
+
+	@ClassRule
+	@Rule
+	public static final LiferayIntegrationTestRule liferayIntegrationTestRule =
+		new LiferayIntegrationTestRule();
+
+	@BeforeClass
+	public static void setUpClass() throws Exception {
+		_bundleId = ModuleFrameworkUtilAdapter.addBundle(
+			PortalIntegrationPointTest.class.getName(), _createBundle());
+
+		ModuleFrameworkUtilAdapter.startBundle(_bundleId);
+	}
+
+	@AfterClass
+	public static void tearDownClass() throws PortalException {
+		if (_bundleId == null) {
+			return;
+		}
+
+		ModuleFrameworkUtilAdapter.stopBundle(_bundleId);
+
+		ModuleFrameworkUtilAdapter.uninstallBundle(_bundleId);
+	}
+
+	@Test
+	public void testPortalIntegrationPointWithServiceProxyFactory() {
+		_displayContextFactory = ServiceProxyFactory.newServiceTrackedInstance(
+			DisplayContextFactory.class, PortalIntegrationPointTest.class,
+			"_displayContextFactory", false);
+
+		Class<?> clazz = _displayContextFactory.getClass();
+
+		Assert.assertEquals(
+			TestDisplayContextFactory.class.getName(), clazz.getName());
+	}
+
+	@Test
+	public void testPortalIntegrationPointWithServiceTracker() {
+		_testPortalIntegrationPointWithServiceTracker(true);
+		_testPortalIntegrationPointWithServiceTracker(false);
+	}
+
+	private static InputStream _createBundle() throws Exception {
+		Class<?> clazz = PortalIntegrationPointTest.class;
```

Comment:
Hi @dantewang @ChiZelin ,

Please simplify the logic of creating bundle.

Now you only have one place to create the bundle on the fly, you do not need to parse everything from class name, or you can do some special hard code if needed.

For example, you do not need a real bnd.bnd file, the only thing you need is to populate a Properties used by Builder.

@dantewang you said "Creating a new interface in test package doesn't work because of package resolving failure from the fake bundle.", I think if you change the logic of resolving here, then you won't have that problem.

Tina